### ISSUE TITLE
Mutex::unlock serialize code after it

### ISSUE THREAD

**Issue by MBkkt** [facebookexperimental/libunifex#449]
I think it's really bad that mutex serialize all waiters to single thread.

I think unlock should be something like this, if it async:
Executor.Schedule(after unlock code);
ExecuteHere(next critical section)

But we don't have async dtor :(, so for unique lock something like:
Executor.Schedule(next critical section)
ExecuteHere(after unlock code)

But now test like this doesn't work:
https://github.com/YACLib/YACLib/blob/main/test/unit/coroutine/async_mutex.cpp#L273

Check Unlock in our library:
https://github.com/YACLib/YACLib/blob/main/include/yaclib/coroutine/async_mutex.hpp

**Comment by MBkkt** [comment]
https://github.com/facebookexperimental/libunifex/blob/main/source/async_mutex.cpp#L40

**Comment by MBkkt** [comment]
@ispeters Yes, but this solution is broke critical section batch execution if I understand it right.

**Comment by MBkkt** [comment]
@ispeters Of course, let me show example

task1:
start execution on thread_pool 1
lock mutex
critical section 1
unlock mutex
continue execution on thread_pool 1

task2:
start execution on thread_pool 2
lock mutex
critical section 2
unlock mutex
continue execution on thread_pool 2

In a lot of cases doesn't matter where will be executed critical section (in the thread_pool 1 or 2).

So if they are executed it the same time (and one lock wait another unlock), would be nice to have ability to execute cs in the batch.
It's also known as flat combining.

**Comment by ispeters** [comment]
@MBkkt Hmm, I think I've misunderstood your initial proposal and I've never heard of "flat combining" before.

Do I understand correctly that you're proposing that `unlock` would become async (it's currently a `void`-returning sync function) so that the current thread can continue with the next task waiting on the mutex, while we schedule resumption of the most-recent-unlocker?

**Comment by MBkkt** [comment]
@ispeters 
Yes I talked about alternative unlock overload which is async.
If unlock async it's possible to use symmetric transfer to continue next critical section in the current os thread, and reschedule code after unlock in some other "executor" (so code after unlock executed in parallel with)

I mentioned flat combining because thought it's known in the facebook
https://github.com/facebook/folly/blob/main/folly/experimental/flat_combining/FlatCombining.h#L33

**Comment by MBkkt** [comment]
@ispeters I agree it's not really related to this issue, maybe I should create the separate issue?

**Comment by ispeters** [comment]
> @ispeters Yes I talked about alternative unlock overload which is async. If unlock async it's possible to use symmetric transfer to continue next critical section in the current os thread, and reschedule code after unlock in some other "executor" (so code after unlock executed in parallel with next critical section)

OK, I get it now.  Neat!

> I mentioned flat combining because thought it's known in the facebook https://github.com/facebook/folly/blob/main/folly/experimental/flat_combining/FlatCombining.h#L33

It's evidently known to some, but I'm only one of tens of thousands of engineers. :P

> maybe I should create the separate issue?

Well, it sounds like there are two ideas here, maybe?  I _thought_ you wanted to reschedule the next waiter before continuing the most-recent-unlocker on the current thread, which should be doable once we merge #527, as I said above; I think that's still something we should do.

Now that I understand the flat combining idea, I think that's a good idea we should explore, too.  Having typed all this out, yeah, please file a new issue and we can use this one to track what I _thought_ you were looking for.

**Comment by MBkkt** [comment]
> Well, it sounds like there are two ideas here, maybe? I thought you wanted to reschedule the next waiter before continuing the most-recent-unlocker on the current thread, which should be doable once we merge #527, as I said above; I think that's still something we should do.

Yes, this is a more important, and it's really nice that it will be fixed.

I created new issue for flat combining  https://github.com/facebookexperimental/libunifex/issues/529
